File: 1056-capacity-to-ship-packages-within-d-days/solution.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int n= weights.length;
        //corner case
        if( days == 0) return -1;

        //find min and max
        int maxi= weights[0];
        int sum =0;
        for( int it : weights){
            if( it > maxi) maxi= it;
            sum+=it;
        }

        //now do binary search and find ans
        int low =maxi, high= sum, ans=-1;

        while( low <= high){
            int mid = low + (high-low)/2;

            if( isPossibleToShipWithMidWeight( mid, weights, days)){
                ans = mid;
                //find more less capacity
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return ans;
    }
    
    public boolean isPossibleToShipWithMidWeight(int wt, int []weights, int d){
         int cntDays = 0;
         int w = wt;
         for( int it : weights){
            if( it <= w){
                w = w-it;
            }else{
                cntDays++;
                w = wt-it;
            }
         }
        cntDays++;
        return cntDays<=d;
   }
}

File: 1056-capacity-to-ship-packages-within-d-days/solution.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function(weights, days) {
    let mini = Math.min(...weights);
    let maxi = Math.max(...weights);

    if(weights.length == days) return maxi;

    let sum = weights.reduce((val, curr)=> val+=curr, 0);

    let s=maxi, e=sum;
    let ans =-1;
    while(s<=e){
        let mid = s + Math.floor((e-s)/2);

        if( isPossibleToShip(mid, weights, days)){
            ans = mid;
            e=mid-1;
        }else s= mid+1;
    }
    return ans;
};

function isPossibleToShip(cap, weights, days){
    let totalWeights = 0;
    let cntDays = 0;
    for(let w  of weights){
        if(totalWeights + w <= cap){
            totalWeights += w;
        }else{
            cntDays++;
            totalWeights = w;
        }
    }

    if(totalWeights != 0) cntDays++;

    return cntDays <= days;
}
